test/s3-resumption/mzworkflows.py

```diff
@@ -0,0 +1,79 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+from os import environ
+from unittest.mock import patch
+
+from materialize.mzcompose import (
+    Localstack,
+    Materialized,
+    Testdrive,
+    Toxiproxy,
+    Workflow,
+)
+
+daemons = [
+    Localstack(services="sts,iam,s3,sqs"),
+    Materialized(environment=["MZ_LOG_FILTER=dataflow::source::s3=trace,info"]),
+    Toxiproxy(),
+]
+
+services = [*daemons, Testdrive()]
+
+#
+# Test the S3 resumption logic by first instructing Toxiproxy to drop a connection
+# after toxiproxy_bytes_allowed have been transfered over it. Then unblock the
+# network and expect full recovery if the interruption has been
+# shorter than the timeout.
+#
+# For different values of bytes_allowed, the following happens:
+# 0 - Connection is dropped immediately
+# 1K - SQS queue and bucket listing are both prevented
+# 2K - SQS and key fetching are both prevented
+# 10K - only key fetching is prevented
+
+
+def workflow_s3_resumption(w: Workflow):
+    w.start_and_wait_for_tcp(services=daemons)
+    w.wait_for_mz()
+
+    for toxiproxy_bytes_allowed in [0, 256, 512, 768, 1024, 1536, 2048, 3072, 5120]:
+        # For small values of toxiproxy_bytes_allowed, we need to allow for CREATE SOURCE to go undisturbed first, otherwise it fails immediately
```

Comment:
Does materialize observe network failures at all when we allow the configuration to finish and then we cut the connection?

Reply:
Yes, it was observing some failures. However, instructing toxyproxy after doing CREATE SOURCE allowed for a window of opportunity for Mz to slurp all the data undisturbed.